### DISCUSSION TITLE
LXQt's portal exits at startup under kwin_wayland (and wayfire)

### DISCUSSION THREAD

**Discussion by ffs-akl** [lxqt/lxqt#2586] [Q&A]
https://github.com/lxqt/xdg-desktop-portal-lxqt/issues/39
After following the issues 39, chromium will use the filechooser of lxqt normally, but after restarting, chromium will use the filechooser of GTK again, and I need to follow the issue 39 again  to get chromium use the filechooser of lxqt.
I  checked by  command systemctl --user status xdg-desktop-portal.service and systemctl --user status xdg-document-portal.service and  systemctl --user status xdg-permission-store.service  before restart them.  all of there service have been active and had been runing.
Do I need to make a script to restart the there service on every time when i log in or put the there service into  ~/.config/systemd/user/ ?

**Comment by tsujan** [comment]
If you've set `GTK_USE_PORTAL`, remove it (just don't set it) and restart the session.

> Do I need to make a script to restart the there service on every time....

No, you don't. Your problem is caused somewhere else, maybe by `GTK_USE_PORTAL`.

**Comment by Bluey26** [comment]
Hi, maybe you need to launch Chromium with a special parameter to make it to use pcmanfm-qt file manager?

Like:

`XDG_CURRENT_DESKTOP=KDE chromium`
(for dolphin)
https://bbs.archlinux.org/viewtopic.php?pid=1762742#p1762742

In firefox, it is enough to create `lxqt-portals.conf` with the specified lines, plus changing the corresponding firefox configuration preferences 

```
widget.use-xdg-desktop-portal.file-picker = 1
widget.use-xdg-desktop-portal.mime-handler = 1
```

**Reply by tsujan** [reply]
I use Chromium as my default browser and haven't encountered any problem after restarting/reboot. Setting `XDG_CURRENT_DESKTOP` to KDE under LXQt can cause troubles.

**Reply by Bluey26** [reply]
Oh, i was not aware of that you were using chromium. 

Then the issue may be somewhere else. I was just thinking about launching chromium using that line, which (i think) will only affect the chromium program. The idea was to change it to `lxqt` or the correct argument to make it work with PCManFM-Qt.

**Reply by tsujan** [reply]
Recently I switched to Chromium under LXQt+LabWC, and I'm happy with it.

The only setting that I changed in Chromium was `chrome://flags/` → `Preferred Ozone platform`; I set it to `Auto`. Since then, Chromium runs under Wayland (not XWayland) and, of course, uses LXQt file dialog.

**Reply by stefonarch** [reply]
Chromium here uses the lxqt-portal too, with
```
~/.config/chromium-flags.conf :

--ozone-platform-hint=auto
```
but also using it under xwayland it still uses lxqt-portal.

**Reply by tsujan** [reply]
> ~/.config/chromium-flags.conf :

Oh, I used it first, and Chromium crashed ;) Using `chrome://flags/` is the safest way to make Chromium run under Wayland.

**Reply by tsujan** [reply]
BTW, one good thing about Chromium is that you can make it use Qt's palette, in the light or dark mode. See its Settings → Appearance → Theme.

**Comment by ffs-akl** [comment]
I've removed GTK_USE_PORTAL=1 and GDK_DEBUG=portals , but chromium still used filechooser of GTK unless  I restarted these three services (xdg-permission-store.service, xdg-document-portal.service,xdg-desktop-portal.service)
I have tried `XDG_CURRENT_DESKTOP=KDE chromium`,  it didn't take effect to chromium.

edited:
>Make ~/.config/xdg-desktop-portal/lxqt-portals.conf as
[preferred]
default=lxqt

Just need to restat these three services (xdg-permission-store.service, xdg-document-portal.service,xdg-desktop-portal.service),  it can get the chromium use filechooser of lxqt without that make ~/.config/xdg-desktop-portal/lxqt-portals.conf

**Reply by tsujan** [reply]
That's strange. Your distro?

**Reply by ffs-akl** [reply]
> That's strange. Your distro?

Arch Linux+LXQt+Kwin_wayland, the wayland file of lxqt is made by @stefonarch

**Reply by tsujan** [reply]
> Arch Linux+LXQt+Kwin_wayland

Oh, I use LXQt+LabWC. Perhaps there's something different about kwin_wayland, some EV that interferes with portals.

**Reply by tsujan** [reply]
@stefonarch, can it be caused by `export XDG_SESSION_DESKTOP="KDE"` in your kwin_wayland script?

**Comment by ffs-akl** [comment]
>@stefonarch, can it be caused by export XDG_SESSION_DESKTOP="KDE" in your kwin_wayland script?

I tried to chang the KDE to LXQt and remove XDG_SESSION_DESKTOP="KDE" ,  the both didn't take effect to chromium.
actually ,I don't feel that XDG_SESSION_DESKTOP="KDE"  has any effect to the LXQt session,  Removing it will have no effect to my lxqt-kwin.

**Comment by ffs-akl** [comment]
> Oh, I use LXQt+LabWC. Perhaps there's something different about kwin_wayland, some EV that interferes with portals.

on my side,  LXQt+Labwc without ~/.config/xdg-desktop-portal/lxqt-portals.conf， it is perfect  for chromium to use filechooser of lxqt

**Reply by tsujan** [reply]
Yes, Chromium works as expected with LXQt+LabWC.

Then, maybe, Chromium itself does something when KWin is running. This is a wild guess; don't count on it.

**Reply by ffs-akl** [reply]
as a temporary solution,I made a script as an autostart program to have the three services restart after logged in every time . But it isn't a good solution.

**Reply by ffs-akl** [reply]
just restart "xdg-desktop-portal.service"  after log in, then the chromium will use the filechooser of lxqt.

**Comment by ffs-akl** [comment]
Is there a way to get these three services( xdg-permission-store.service, xdg-document-portal.service,xdg-desktop-portal.service) to start after logged in?

**Reply by stefonarch** [reply]
In`~/.config/labwc/autostart` I use `systemctl --user start xdg-desktop-portal-wlr.service  > /dev/null 2>1 &`

**Reply by ffs-akl** [reply]
do you mean that i need to start  xdg-desktop-portal-lxqt.service in lxqt-kwin by autostart?
beacause i am in lxqt-kwin, i only installed the xdg-desktop-portal-lxqt.

**Reply by ffs-akl** [reply]
and  just checked with` systemctl --user status xdg-desktop-portal-lxqt`,  output message is `Unit xdg-desktop-portal-lxqt.service could not be found`

**Reply by stefonarch** [reply]
`xdg-desktop-portal-lxqt` should always start, this example was for `wlr`. In kwin there's only one autostart for both x11 and wayland, so you need to set it there for the services you want:

` systemctl --user start  xdg-permission-store.service` works here, no idea what it does though. The other 2 portals are running here without any particular setup.

**Reply by ffs-akl** [reply]
on my side,  the there service are also running here without any particular setup. It's just that I think they started too early, and I wish they started a little later

**Comment by bbigam** [comment]
I don't know whether or not this is related, but on openSUSE Tumbleweed, xdg-desktop-portal does not automatically pick up xdg-desktop-portal-lxqt on login. I had to add `/usr/libexec/xdg-desktop-portal --replace` to my autostart apps, and then it works consistently.

It's been this way for a long while and through various reinstalls and settings resets, with both Openbox and Xfwm4. I reported the bug to openSUSE (https://bugzilla.opensuse.org/show_bug.cgi?id=1226394) over a month ago, but no one has responded to it.

**Reply by ffs-akl** [reply]
It's the same issue about xdg-desktop-portal,    a script to restart the xdg-desktop-portal  can fix the issue.

**Comment by tsujan** [comment]
At last, I found time to test kwin_wayland+LXQt comprehensively and compare it with labwc+LXQt yesterday. Previously kwin_wayland seemed too buggy, but I found a way of avoiding its bugs.

Everything was OK, until I found out that Chromium showed the GTK file dialog, instead of LXQt's. Worse than that: Firefox didn't show anything! In other words, I encountered the problem discussed here.

In my log, I found this line:

```
dbus-:1.2-org.freedesktop.impl.portal.desktop.lxqt@0.service: Failed with result 'exit-code'
```

It shows that LXQt's portal was started but failed at the login.

Of course, it's possible to use an auto-start scrip to restart the user services, but I think the problem is in kwin_wayland (although I can't prove that yet).

I have Manjaro Testing, and they haven't added kwin-6.1.X to its repository from Arch — apparently, they believe that KDE 6.1.X is too buggy — so, my kwin's version is still 6.0.5-3. Since @stefonarch doesn't see the problem, I think the problem may have been fixed in kwin-6.1.X from Arch. On the other hand, if kwin ≥ 6.1.X comes here and the problem persists, the cause should be elsewhere.

**Reply by ffs-akl** [reply]
By the time I found out about this and started this discussion here, I had already used KWIN_WAYLAND 6.1.2, so I think the problem should be elsewhere

**Reply by tsujan** [reply]
@knm100
Thanks for the info.

Our problem is that @stefonarch doesn't encounter the problem ;) And we don't know why. Some setting should be different for him: a needle in the haystack.

**Comment by tsujan** [comment]
The interesting thing is that exactly the same problem happened to me under Wayfire. Apparently, only Labwc is immune to it.

**Reply by ffs-akl** [reply]
Are the sway and hyprland immune to it?

**Reply by stefonarch** [reply]
I've no issues anywhere with the portal exiting.

**Reply by tsujan** [reply]
> I've no issues anywhere with the portal exiting.

You should tell us why. We're the majority ;)

**Reply by Bluey26** [reply]
> > I've no issues anywhere with the portal exiting.
> 
> You should tell us why. We're the majority ;)

I think there is something else going on in here. I just have installed chromium to give some light into this. I have wayfire and sway installed and opened chromium in `labwc`, and simulated a save file into it. LXQt file dialog (pcmanf-qt like) appeared (desired behavior).

I did the same in `wayfire` and `sway`, and in both they work, i have not seen any crash in here.I have not ran any of the commands mentioned previously in here, it was pretty straightforward : installed `chromium` package form archlinux repositories, and then i opened it.

For portals, i already had the file configured and set up for firefox.

Mine looks like this:

`~/.config/xdg-desktop-portal/portals.conf`

```
[preferred]
default=gtk;wlr
org.freedesktop.impl.portal.FileChooser=lxqt
```

I wonder if the OS is a key factor in here, for whatever reason

**Reply by tsujan** [reply]
> I wonder if the OS is a key factor in here

I don't think so. I have Manjaro, which is 99% Arch.

> Mine looks like this:

That's different from mine. I have `~/.config/xdg-desktop-portal/lxqt-portals.conf`:

```
[preferred]
default=lxqt
```

Why `default=gtk;wlr`? Do you have a link that explains the contents of your file?

**Reply by tsujan** [reply]
@stefonarch 
What's your portal config file (its name and contents)?

**Reply by Bluey26** [reply]
> Why `default=gtk;wlr`? Do you have a link that explains the contents of your file?

I think it was taken from this discussion: https://github.com/labwc/labwc/discussions/1503

I do not remember exactly how i ended up with that file. It was copied by me from the older lxqt+labwc setup after i changed my hard drive.

I left it as that because i was trying to make a mpv front-end (smplayer) to follow the wayland idle-inhibit protocol and i have stumbled with that topic.

After a few more workarounds, smplayer does follow my idle-inhibit (an experimental build, but it works) so i did not touched anything since then.

**UPDATE**

The semicolon part seems to be from here:

https://man.archlinux.org/man/extra/xdg-desktop-portal/portals.conf.5.en

>org.freedesktop.impl.portal.* (string)

> One of the valid portal interface implementations exposed by xdg-desktop-portal.
> 
> Each key in the group contains a semi-colon separated list of portal backend implementation, to be searched for an implementation of the requested interface, in the same order as specified in the configuration file. Additionally, the special values none and * can be used:

I may be missing some more information, i usually ask things in #labwc IRC room so i dont remember if this was suggested in there or not.

**Reply by stefonarch** [reply]
> @stefonarch What's your portal config file (its name and contents)?

no file  in that directory, as said.

**Reply by tsujan** [reply]
Using of portal under Wayland shouldn't be that complex.

I removed my config file and restarted the user service. There was no change: everything is OK under Labwc, while LXQt portal exits at startup under kwin_wayland. And like before, restarting of the user service brings it back under kwin_wayland (@knm100's workaround).

We need to find the hidden factor!

**Reply by tsujan** [reply]
To be thorough, I add that I have `KillUserProcesses=yes` in `/etc/systemd/logind.conf`.

**Reply by Bluey26** [reply]
> To be thorough, I add that I have `KillUserProcesses=yes` in `/etc/systemd/logind.conf`.

`#KillUserProcesses=no` in here, which means i have the default value (`KillUserProcesses=no`)

**Reply by tsujan** [reply]
> #KillUserProcesses=no in here

@knm100, @stefonarch, and for you?

**Reply by stefonarch** [reply]
all commented out in that file

**Reply by tsujan** [reply]
OK, that may explain it.

IMO, `KillUserProcesses=no` is a bad default. With it, the user processes won't be stopped under environments that don't implement a mechanism for ending the running apps.

Anyway, the question still remains why LXQt's portal starts normally under Labwc with `KillUserProcesses=yes` but not under some other Wayland compositors.

**Reply by stefonarch** [reply]
Just saw that in debian gtk-portal was used, removed it. No config file.  It says to use it :)

```
systemctl --user status xdg-desktop-portal.service 

set 01 21:56:43 archlinux xdg-desktop-por[1657328]: No skeleton to export
set 01 21:56:43 archlinux xdg-desktop-por[1657328]: Choosing lxqt.portal for org.freedesktop.impl.portal.FileChooser via the deprecated UseIn key
set 01 21:56:43 archlinux xdg-desktop-por[1657328]: The preferred method to match portal implementations to desktop environments is to use the portals.conf(5) configuration file
set 01 21:56:43 archlinux xdg-desktop-por[1657328]: Choosing wlr.portal for org.freedesktop.impl.portal.ScreenCast via the deprecated UseIn key
```

**Reply by ffs-akl** [reply]
@tsujan   all commented out in /etc/systemd/logind.conf

**Reply by tsujan** [reply]
> all commented out in /etc/systemd/logind.conf

Then `KillUserProcesses` can't be related to this.

**Reply by ffs-akl** [reply]
Just an infomation. I installed the edge to my lxqt+labwc,  then the chromium back to use the gtk filechooser, and edge used the gtk filechooser too.
 running command` systemctl --user status xdg-desktop-portal.service` , showed this service was actived and running.   
after restart this service,   the chromium and edge revert back to use lxqt filechooser.  
Rebooted and tested, the chromium and edge still used  the lxqt filechooser.     
Edited :
Removed the chromium,  the edge still used the lxqt filechooser.

**Comment by tsujan** [comment]
The problem was solved here by replacing SDDM with greetd+tuigreet. I got rid of of SDDM's crashes and freezes too.

But setting up greetd may need some expert knowledge.

**Reply by ffs-akl** [reply]
Sounds reasonable,  I am good at setting up greetd with labwc.  I will try it with kwin_wayland.   it is first time to run greetd by kwin_wayland.  do you mind shareing your config file fo greetd with kwin_wayland?

**Reply by ffs-akl** [reply]
> it is first time to run greetd by kwin_wayland.  do you mind shareing your config file fo greetd with kwin_wayland?

Please ignore it.
tuigreet has nothing to do with compositor, so there should be no information in your config file about starting a wayland greeter with kwin_wayland either.

**Reply by tsujan** [reply]
I learned it from @stefonarch and added what I needed.

My `/etc/greetd/config.toml`:

```toml
[terminal]
vt = 1

[default_session]
command = "tuigreet --remember -i -t --remember-session --cmd startlxqtwayland"

user = "greeter"
```

And this is my `/etc/pam.d/greetd` (because I use NetworkManager with gnome-keyring):

```
#%PAM-1.0

auth       required     pam_securetty.so
auth       requisite    pam_nologin.so
auth       include      system-local-login
auth       optional     pam_gnome_keyring.so
account    include      system-local-login
session    include      system-local-login
session    optional     pam_gnome_keyring.so auto_start
```

The most important thing — that was confusing to me at first — was that the `command` in `/etc/greetd/config.toml` did *not* work properly by itself. Then I realized that, instead, I should press `F3` and choose my Wayland session *explicitly*, once for all. Of course, the session could be changed later by pressing `F3` again.

The reason is that a proper Wayland command needs the environment variable `XDG_SESSION_TYPE` to be set to `wayland`, but greetd doesn't accept EVs. However, that will happen automatically if a Wayland session is chosen by pressing `F3`.

The advantage of greetd+tuigreet over other greeters is that it doesn't use X11 or Wayland for logging in. Its look is elementary, but it does its job properly. My reason for using it wasn't related to the problem discussed here. SDDM crashed or made the whole system freeze after logging out several times. That happened for several months and didn't let me switch safely between the three Wayland compositors I had.

**Reply by tsujan** [reply]
> tuigreet has nothing to do with compositor

That's right. And to me, it's an advantage (as explained above).

**Reply by ffs-akl** [reply]
> I learned it from @stefonarch and added what I needed.

that is why @stefonarch didn't encounter this issue.

**Reply by ffs-akl** [reply]
Thank you for your shareing

**Reply by tsujan** [reply]
> that is why @stefonarch didn't encounter this issue.

Maybe, but I'm not sure. Systems are so complex that a single problem may have multiple causes.

Anyway, we'd found a simple workaround for the problem discussed here. My regret was that I wasted my time with SDDM, hoping that it might work reliably at some point, but it never did.

**Reply by stefonarch** [reply]
It was fine also before with sddm here ;)

**Comment by reigneil** [comment]
any progress with this issue?
as @bbigam reported this on openSUSE it looks like they finally fix it?
https://build.opensuse.org/request/show/1244264

**Reply by tsujan** [reply]
> … it looks like they finally fix it?

It's another workaround, not a fix.

**Reply by reigneil** [reply]
for weird reason i fix my issue by just renaming `lxqt-portals.conf` to just `portals.conf`, restart and now i saw xdg-desktop-portal-lxqt running in background.
